src/test/java/StringManipulationTest.java

```diff
@@ -19,131 +22,206 @@ public void tearDown() {
     }
+
+    // Test case for extracting substrings from a null string
     @Test
     public void testGeSubStrings4() {
-        fail("Not yet implemented");
+        // Set up the initial string to null
+        manipulatedstring.setString(null);
+        assertThrows(NullPointerException.class, () -> manipulatedstring.getSubStrings(2, 5));
     }
+
+    // Test case for extracting substrings with invalid startWord and endWord values
     @Test
     public void testGeSubStrings5() {
-        fail("Not yet implemented");
+        // Set up the initial string
+        manipulatedstring.setString("Hello!");
+        assertThrows(IllegalArgumentException.class, () -> manipulatedstring.getSubStrings(-1, 0));
     }
+
+    // Test case to verify extracting a single substring
     @Test
     public void testGeSubStrings6() {
-        fail("Not yet implemented");
+        // Set up the initial string
+        manipulatedstring.setString("0 1");
+        String[] sString = manipulatedstring.getSubStrings(1, 2);
+        assertEquals(sString[0], "0");
     }
 
+    // DO NOT EDIT
     @Test
-    public void testRestoreString1()
-    {
+    public void testRestoreString1() {
+        // Set up the initial string
         manipulatedstring.setString("art");
-        int [] array;
-        array=new int[]{1,0,2};
+        int[] array;
+        array = new int[] { 1, 0, 2 };
         String restoreString = manipulatedstring.restoreString(array);
         assertEquals(restoreString, "rat");
     }
 
+    // Test case for restoring an empty string
     @Test
-    public void testRestoreString2()
-    {
-        fail("Not yet implemented");
-
+    public void testRestoreString2() {
+        // Set up the initial string
+        manipulatedstring.setString("");
+        int[] indices = {};
+        String result = manipulatedstring.restoreString(indices);
+        assertEquals("", result);
     }
 
+    // Test case for restoring a string with an out-of-bounds index in the indices
+    // array
     @Test
-    public void testRestoreString3()
-    {
-        fail("Not yet implemented");
-
+    public void testRestoreString3() {
+        // Set up the initial string
+        manipulatedstring.setString("abc");
+        // seting up indices for shuffling, this one goes out of bounds
+        int[] indices = { 1, 2, 4 };
+        assertThrows(IndexOutOfBoundsException.class, () -> manipulatedstring.restoreString(indices));
     }
 
+    // Test case for restoring a null string
     @Test
-    public void testRestoreString4()
-    {
-        fail("Not yet implemented");
-
+    public void testRestoreString4() {
+        // setting up null string
+        manipulatedstring.setString(null);
+        // setting indicies
+        int[] indices = { 0, 1, 2 };
+        // throwing an exception for null string
+        assertThrows(IllegalArgumentException.class, () -> manipulatedstring.restoreString(indices));
```

Comment:
This test should throw a null pointer exception if it is testing for a null string